ibis/backends/dask/execution/generic.py

```diff
@@ -0,0 +1,869 @@
+"""Execution rules for generic ibis operations."""
+
+import collections
+import datetime
+import decimal
+import functools
+import numbers
+import operator
+from typing import Optional
+
+import dask.array as da
+import dask.dataframe as dd
+import numpy as np
+import pandas as pd
+from dask.dataframe.groupby import DataFrameGroupBy, SeriesGroupBy
+from pandas import isnull, to_datetime
+from pandas.api.types import DatetimeTZDtype
+
+import ibis.common.exceptions as com
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+import ibis.expr.types as ir
+from ibis.backends.pandas.core import (
+    boolean_types,
+    execute,
+    floating_types,
+    integer_types,
+    numeric_types,
+    scalar_types,
+    simple_types,
+)
+from ibis.backends.pandas.execution import constants
+from ibis.backends.pandas.execution.generic import execute_node
+from ibis.expr.scope import Scope
+from ibis.expr.timecontext import TIME_COL
+from ibis.expr.typing import TimeContext
+
+# from .. import aggcontext as agg_ctx
+from ..client import DaskClient, DaskTable
+
+
+@execute_node.register(ops.Limit, dd.DataFrame, integer_types, integer_types)
+def execute_limit_frame(op, data, nrows, offset, **kwargs):
+    return data.loc[offset : offset + nrows]
+
+
+@execute_node.register(ops.Cast, SeriesGroupBy, dt.DataType)
+def execute_cast_series_group_by(op, data, type, **kwargs):
+    result = execute_cast_series_generic(op, data.obj, type, **kwargs)
+    return result.groupby(data.grouper.groupings)
+
+
+@execute_node.register(ops.Cast, dd.Series, dt.DataType)
+def execute_cast_series_generic(op, data, type, **kwargs):
+    return data.astype(constants.IBIS_TYPE_TO_PANDAS_TYPE[type])
+
+
+@execute_node.register(ops.Cast, dd.Series, dt.Array)
+def execute_cast_series_array(op, data, type, **kwargs):
+    value_type = type.value_type
+    numpy_type = constants.IBIS_TYPE_TO_PANDAS_TYPE.get(value_type, None)
+    if numpy_type is None:
+        raise ValueError(
+            'Array value type must be a primitive type '
+            '(e.g., number, string, or timestamp)'
+        )
+    return data.map(
+        lambda array, numpy_type=numpy_type: list(map(numpy_type, array))
+    )
+
+
+@execute_node.register(ops.Cast, dd.Series, dt.Timestamp)
+def execute_cast_series_timestamp(op, data, type, **kwargs):
+    arg = op.arg
+    from_type = arg.type()
+
+    if from_type.equals(type):  # noop cast
+        return data
+
+    tz = type.timezone
+
+    if isinstance(from_type, (dt.Timestamp, dt.Date)):
+        return data.astype(
+            'M8[ns]' if tz is None else DatetimeTZDtype('ns', tz)
```

Comment:
Is this not handled by IBIS_TYE_TO_PANDAS_TYPE?

Reply:
I don't believe that keeps the timezone info on there.

Reply:
Looks like that would be xfailing a lot of tests. Is that what you expected?

```
FAILED ibis/backends/dask/tests/execution/test_cast.py::test_cast_timestamp_column[plain_datetimes_naive-to3-datetime64[ns, America/Los_Angeles]] - TypeError: Don't know how to cast timestamp to timestamp('America/Los_Angeles')
FAILED ibis/backends/dask/tests/execution/test_cast.py::test_cast_timestamp_column[plain_datetimes_naive-timestamp('America/Los_Angeles')-datetime64[ns, America/Los_Angeles]] - TypeError: Don't know how to cast timestamp to timestamp...
FAILED ibis/backends/dask/tests/execution/test_cast.py::test_cast_timestamp_column[plain_datetimes_ny-to3-datetime64[ns, America/Los_Angeles]] - TypeError: Don't know how to cast timestamp('America/New_York') to timestamp('America/Lo...
FAILED ibis/backends/dask/tests/execution/test_cast.py::test_cast_timestamp_column[plain_datetimes_ny-timestamp('America/Los_Angeles')-datetime64[ns, America/Los_Angeles]] - TypeError: Don't know how to cast timestamp('America/New_Yo...
FAILED ibis/backends/dask/tests/execution/test_cast.py::test_cast_timestamp_column[plain_datetimes_utc-to3-datetime64[ns, America/Los_Angeles]] - TypeError: Don't know how to cast timestamp('UTC') to timestamp('America/Los_Angeles')
FAILED ibis/backends/dask/tests/execution/test_cast.py::test_cast_timestamp_column[plain_datetimes_utc-timestamp('America/Los_Angeles')-datetime64[ns, America/Los_Angeles]] - TypeError: Don't know how to cast timestamp('UTC') to time...
FAILED ibis/backends/dask/tests/execution/test_temporal.py::test_times_ops_with_tz[plain_datetimes_utc-US/Eastern-zeros] - TypeError: Don't know how to cast timestamp('UTC') to timestamp('US/Eastern')
FAILED ibis/backends/dask/tests/execution/test_temporal.py::test_times_ops_with_tz[plain_datetimes_utc-None-ones] - TypeError: Don't know how to cast timestamp('UTC') to timestamp
FAILED ibis/backends/dask/tests/execution/test_temporal.py::test_times_ops_with_tz[plain_datetimes_naive-US/Eastern-zeros] - TypeError: Don't know how to cast timestamp to timestamp('US/Eastern')
FAILED ibis/backends/dask/tests/execution/test_temporal.py::test_times_ops_with_tz[plain_datetimes_naive-UTC-ones] - TypeError: Don't know how to cast timestamp to timestamp('UTC')
```

Reply:
hmm what does the pandas backend do here? (note not really concerned about this right now, just add a TODO and can followup later)

Reply:
Basically the same code:

```python
    if isinstance(from_type, (dt.Timestamp, dt.Date)):
        return data.astype(
            'M8[ns]' if tz is None else DatetimeTZDtype('ns', tz)
        )
```
https://github.com/ibis-project/ibis/blob/400b079bea2660fa23ca2ac8faadd15da523ce9d/ibis%2Fbackends%2Fpandas%2Fexecution%2Fgeneric.py#L123-L126

Reply:
Yeah this bit smells a bit. Can you create an issue to track this for both Pandas and Dask?